core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java

```diff
@@ -124,6 +140,103 @@ public StructuredRecord getCurrentValue() throws IOException, InterruptedExcepti
     }
 
     @Override
+    public boolean nextKeyValue() throws IOException, InterruptedException {
+      return delegate.nextKeyValue();
+    }
+
+    @Override
+    public float getProgress() throws IOException, InterruptedException {
+      return delegate.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      delegate.close();
+    }
+  }
+
+  private static class TrackingAvroRecordReader extends RecordReader<NullWritable, StructuredRecord> {
```

Comment:
these classes have a lot of code that is exactly the same. This is usually a sign that they should have a common parent. Something like:

```
private abstract static class TrackingReader<K, V> extends RecordReader<NullWritable, StructuredRecord> {
  private final RecordReader<K, V> delegate;
  protected final Schema schema;
  protected final String pathField;
  protected final String path;

  protected TrackingReader(RecordReader<K, V> delegate, @Nullable String pathField, String path) {
    ..
  }

  // common methods (initialize, getProcess, close, getCurrentKey, nextKeyValue) here
  
  protected abstract StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder);

  public StructuredRecord getCurrentValue() {
    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
    if (pathField != null) {
      recordBuilder.set(pathField, path);
    }
    return getCurrentValue(recordBuilder);
  }
}
```

subclasses would then only be responsible for filling out the fields in the builder.